### PR TITLE
ENG-14941, discover dangling PBD files on nodes that are assigned dif…

### PR DESCRIPTION
**Pull request by luyangco**
…ferent set of partitions. Close and delete the files once all reader are passed the end sequence number. Also fix some corner cases of export stats miscalculation.

Change-Id: I757a91065b3db9584c36069beba25e3e6beeceec

### CODE REVIEW

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 1]
```diff
@@ -978,6 +970,8 @@ private synchronized void pollImpl(SettableFuture<AckingContainer> fut) {
             } finally {
                 //Try hard not to leak memory
                 for (StreamBlock sb : blocksToDelete) {
+                    int tuplesDeleted = m_gapTracker.truncate(sb.lastSequenceNumber());
+                    m_tuplesPending.addAndGet(-tuplesDeleted);
```
Comment (rdykiel) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: Use decrementAndGet? This comment was already made, I think.

File: src/frontend/org/voltdb/export/ExportGeneration.java [hunk 2]
```diff
@@ -172,7 +166,17 @@ void initialize(HostMessenger messenger,
...
             }
         }
-        return foundPartitions;
+
+        // Count unique partitions only
+        Set<Integer> allLocalPartitions = localPartitionsToSites.stream()
+                .map(p -> p.getFirst())
+                .collect(Collectors.toSet());
+        Set<Integer> onDIskPartitionsSet = new HashSet<Integer>(onDiskPartitions);
+        onDIskPartitionsSet.removeAll(allLocalPartitions);
```
Comment (rdykiel) [inline src/frontend/org/voltdb/export/ExportGeneration.java]: I have a problem understanding this: we only create mailbox for partition that is NOT local? Can you explain?

File: src/frontend/org/voltdb/export/ExportGeneration.java [hunk 3]
```diff
@@ -226,121 +228,152 @@ private void updateDataSources( List<String> exportSignatures) {
...
+        // in case it couldn't find local mailbox to send back response, add local mailbox to the list first.
         for (Integer partition : localPartitions) {
             updateAckMailboxes(partition, null);
         }
         // Update latest replica list to each data source.
         updateReplicaList(messenger, localPartitions);
     }
 
+    // Auto reply a response when the requested stream is no longer exists
+    private void sendDummyTakeMastershipMessage(long sourceHsid, long requestId, int partitionId, byte[] signatureBytes) {
```
Comment (rdykiel) [inline src/frontend/org/voltdb/export/ExportGeneration.java]: IMO this method should be called "sendDummyTakeMastershipResponse"

File: src/frontend/org/voltdb/export/ExportDataSource.java (hunk 1 above)
Comment (luyangco) [reply src/frontend/org/voltdb/export/ExportDataSource.java]: decrementAndGet() doesn't take parameter, it decrements the counter by one each call.

File: src/frontend/org/voltdb/export/ExportGeneration.java (hunk 2 above)
Comment (luyangco) [reply src/frontend/org/voltdb/export/ExportGeneration.java]: If you take a step back to ExportGeneration.initialize(), after creating mailbox for NOT local partitions, export also creates mailbox for local partitions.

File: src/frontend/org/voltdb/export/ExportGeneration.java (hunk 3 above)
Comment (luyangco) [reply src/frontend/org/voltdb/export/ExportGeneration.java]: I'll change the name.

**Pull request merged**